substratest/factory.py

```diff
@@ -157,6 +185,20 @@ class TraintupleSpec(_Spec):
     rank: int = None
 
 
+@dataclasses.dataclass
+class CompositeTraintupleSpec(_Spec):
+    algo_key: str
+    objective_key: str
+    data_manager_key: str
+    train_data_sample_keys: str
+    in_head_model_key: str
+    in_trunk_model_key: str
+    tag: str
+    compute_plan_id: str
+    out_trunk_model_permissions: typing.Dict
+    rank: int = None
+
+
```

Comment:
It could be nice to have a  `_TraintupleSpec` as you have done for `_Algo`

Reply:
Yes I thought about this and I'm not 100% sure about what we should do.
For the `Algo` it makes more sense as all the fields are common for an `Algo` and a `CompositeAlgo`.
Thi is not true for `Traintuple` / `CompositeTraintuple`. I would wait a refactoring and finishing the implementation of the `AggregateTraintuple` before factorizing code.